scripts/validate_rst_title_capitalization.py

```diff
@@ -0,0 +1,306 @@
+
+
+def follow_capitalization_convention(title: str) -> bool:
+    """
+    Algorithm to determine if a heading follows the capitalization convention
+
+    This method returns true if the title follows the convention
+    and false if it does not
+
+    Parameters
+    ----------
+    title : str
+        Heading string to validate
+
+    Returns
+    -------
+    bool
+        True if capitalization is correct, False if not
+
+    """
+
+    # split with delimiters comma, semicolon and space, parentheses, colon, slashes
+    word_list = re.split(r"[;,-/():\s]\s*", title)
+
+    # Edge Case: First word is an empty string
+    if len(word_list[0]) == 0:
+        return False
+
+    # Dealing with the first word of the title
+    if word_list[0] not in CAPITALIZATION_EXCEPTIONS:
+        # word is not in CAPITALIZATION_EXCEPTIONS but has different capitalization
+        if word_list[0].lower() in CAPITALIZATION_EXCEPTIONS_LOWER:
+            return False
+        # First letter of first word must be uppercase
+        if not word_list[0][0].isupper():
+            return False
+        # Remaining letters of first word must not be uppercase
+        for j in range(1, len(word_list[0])):
+            if word_list[0][j].isupper():
+                return False
+
+    # Remaining letters must not be uppercase letters
+    for i in range(1, len(word_list)):
+        if word_list[i] not in CAPITALIZATION_EXCEPTIONS:
+            # word is not in CAPITALIZATION_EXCEPTIONS but has different capitalization
+            if word_list[i].lower() in CAPITALIZATION_EXCEPTIONS_LOWER:
+                return False
+            # Remaining letters must not be uppercase
+            for j in range(len(word_list[i])):
+                if word_list[i][j].isupper():
+                    return False
+
+    # Returning True if the heading follows the capitalization convention
+    return True
+
+
+def findTitles(rst_file: str) -> Generator[List[str], List[int], None]:
+    """
+    Algorithm to identify particular text that should be considered headings in an
+    RST file
+
+    See <https://thomas-cokelaer.info/tutorials/sphinx/rest_syntax.html> for details
+    on what constitutes a string as a heading in RST
+
+    Parameters
+    ----------
+    rst_file : str
+        RST file to scan through for headings
+
+    Returns
+    -------
+    title_list : List[str]
+        A list of heading strings found in the document tree
+
+    line_number_list : List[int]
+        The corresponding line numbers of the headings in title_list
+
+    """
+
+    # title_list is the list of headings that is encountered in the doctree
+    title_list: List[str] = []
+
+    # List of line numbers that corresponding headings in title_list can be found at
+    line_number_list: List[int] = []
+
+    # Open and read the .rst file and store the string of data into input
+    f = open(rst_file, "r")
+    input = f.read().split("\n")
+    f.close()
```

Comment:
Can you use a contex manager here?

I believe that

```python
with open(rst_file, "r") as file_obj:
    lines = file_obj.read().split("\n")
```

Should do the trick here.

side note, ```input``` is a built-in python function, please try to avoid using it as a variable name